### ISSUE TITLE
Improve Travis speed: switch back to container

### ISSUE THREAD

**Issue by henn** [CCI-MOC/hil#834]
If we switch back to Travis's container infrastructure, our CI will [start faster and possibly have more CPU resources](https://docs.travis-ci.com/user/reference/precise/#virtualization-environments).

The downside is that we lose sudo ability.

From what I can think of, the biggest holdback would be apache, since it's a system-installed application and we need to mess with its config file.

Note that we can install apache in the container using the [apt stanza](https://docs.travis-ci.com/user/installing-dependencies/) in .travis.yml, as apache is in their [package whitelist](https://github.com/travis-ci/apt-package-whitelist/blob/f343403ebbfeec2bc315d8d83b1d10b8a49502ea/ubuntu-trusty).

This could be more of a "nice to have", as the effort to do this might not be worth the time.

**Comment by zenhack** [comment]
We'd still need to configure apache, which may be harder to do without root.
Can we maybe split it off into it's own section of the build matrix, and use containers for most stuff, only having the stuff that needs root in the vm builds? Or does travis not let you mix & match?

**Comment by henn** [comment]
From [what I can tell](https://docs.travis-ci.com/user/reference/trusty/), one only gets a single .travis.yml, which has a single global "sudo" statement.

Another option would be to add another CI service like CircleCI, though this would increase the points of failure.

**Comment by zenhack** [comment]
Quoting Jason Hennessey (2017-07-28 18:07:16)

>    From [1]what I can tell, one only gets a single .travis.yml, which has
>    a single global "sudo" statement.

Darn.

>    Another option would be to add another CI service like CircleCI, though
>    this would increase the points of failure.

Not worth the trouble, I think.

**Comment by henn** [comment]
It turns out I was wrong -- we **can** mix sudo and non-sudo travis jobs!

Here was the amazing response Travis support sent back to me:

> It is possible to configure different jobs mixing
> sudo: required and false in a single project! To achieve this, you can
> use the Build Matrix feature and define different jobs for sqlite,
> postgres and a third one in which `sudo: required` is explicitly
> configured.You can do so by adding a configuration similar to the
> following example in your .travis.yml file:language: python
> 
```yaml
matrix:
  include:
  - sudo: required
    dist: trusty
    python: "2.7"
    env: SUDO_ENABLED=yes DB=other
  - sudo: false
    dist: trusty
    python: "2.7"
    env: DB=sqlite
  - sudo: false
    dist: trusty
    python: "2.7"
    env: DB=postgres   

install:
  - if [[ "$SUDO_ENABLED" == "yes" ]]; then echo "your special Apache config commands here!"; fi 
```
> Then, as shown in the `install` phase above, you can use conditional
> statements to run any specific commands for the Apache configuration
> only in the sudo-enabled job, and avoid running them through the other
> jobs if they're not needed. Do you think that a setup similar to this
> would be an option for you and your team?As a bit more information, in
> this case, as the sudo: required job is defined as one more job in the
> matrix, these three jobs will run in parallel. If you prefer to run
> these jobs sequentially I'd recommend taking a look at our Build
> Stages beta feature, that will allow you to define different stages in
> your build that will run only if the previous stages were successful.
> For example, you can do this to only run a stage (job) for Apache if
> the previous jobs succeeded.

BTW - I still think we'll get more performance bang for our buck by doing #833 first.

**Comment by naved001** [comment]
@zenhack Do you think it's worth switching to a container in the future? If not, we can close this.

**Comment by zenhack** [comment]
This might be worth it at some point, but it strikes me as relatively low priority. I'd say keep it open though.

The slowest run we have right now though is the apache run, which we can't do in a container, so I dunno.
